File: LeetCode/JavaScript/B18-328-Odd Even Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var oddEvenList = function (head) {
  if (!head) {
    return head;
  }
  let odd = head,
    even = head.next;
  let even_head = head.next;

  while (!!even && !!even.next) {
    [odd.next, even.next] = [odd.next.next, even.next.next];
    [odd, even] = [odd.next, even.next];
  }
  odd.next = even_head;
  return head;
};

File: LeetCode/5회차/B18-328-Odd Even Linked List.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        odd_root = odd = ListNode()
        even_root = even = ListNode()

        while head:
            odd.next = head
            odd, head = odd.next, head.next
            if not head:
                break
            even.next = head
            even, head = even.next, head.next

        odd.next, even.next = even_root.next, None
        return odd_root.next
